ContactMechanics/PipelineFunction.py

```diff
@@ -0,0 +1,327 @@
+    try:
+        # Reset plastic displacement if this is a plastic calculation. We need to do this because the individual steps
+        # are not in order, i.e. the contact is not continuously formed or lifted. Each calculation needs to compute
+        # a fresh plastic displacement.
+        topography.plastic_displ = np.zeros_like(topography.plastic_displ)
+    except AttributeError:
+        pass
+
+    # Get the profile as a numpy array
+    heights = topography.heights()
+
+    # Find max, min and mean heights
+    top = np.max(heights)
+    middle = np.mean(heights)
+    bot = np.min(heights)
+
+    if history is None:
+        step = 0
+    else:
+        mean_displacements, mean_gaps, mean_pressures, total_contact_areas, converged = history
+        step = len(mean_displacements)
+
+    if step == 0:
+        mean_displacement = -middle
+    elif step == 1:
+        mean_displacement = -top + 0.01 * (top - middle)
+    else:
+        # Intermediate sort by area
+        sorted_disp, sorted_area = np.transpose(
+            sorted(zip(mean_displacements, total_contact_areas), key=lambda x: x[1]))
+
+        ref_area = np.log10(np.array(sorted_area + 1 / np.prod(topography.nb_grid_pts)))
+        darea = np.append(ref_area[1:] - ref_area[:-1], -ref_area[-1])
+        i = np.argmax(darea)
+        if i == step - 1:
+            mean_displacement = bot + 2 * (sorted_disp[-1] - bot)
+        else:
+            mean_displacement = (sorted_disp[i] + sorted_disp[i + 1]) / 2
+
+    return _contact_calculation(system, offset=mean_displacement, pentol=pentol, maxiter=maxiter, history=history,
+                                optimizer_kwargs=optimizer_kwargs)
+
+
+def contact_mechanics(self, substrate=None, nsteps=None, offsets=None, pressures=None, hardness=None, maxiter=100,
+                      callback=None, optimizer_kwargs={}):
+    """
+    Carry out an automated contact mechanics calculations.
+
+    Parameters
+    ----------
+    self : :obj:`SurfaceTopography.UniformTopographyInterface`
+        Topography on which to carry out the contact calculation.
+    substrate : str, optional
+        Specifies whether substrate should be 'periodic' or 'nonperiodic'. If
+        set to None, it will be chosen according to whether the topography is
+        periodic or nonperiodic.
+        (Default: None)
+    nsteps : int, optional
+        Number of contact steps. (Default: 10)
+    offsets : list of floats, optional
+        List with offsets. Can only be set if `nsteps` and `pressures` is
+        set to None. (Default: None)
+    pressures : list of floats, optional
+        List with pressures in units of E*. Can only be set if `nsteps` and
+        `offsets` is set to None. (Default: None)
+    hardness : float, optional
+        Hardness in units of E*. Calculation is fully elastic if set to None.
+        (Default: None)
+    maxiter : int, optional
+        Maximum number of interations. (Default: 100)
+    callback : func, optional
```

Comment:
can you be more explicit that this is called at each penetration and force step after the minimization ? It can be confused the callback called at each iteration of the optmization.

Other suggested names: dump_results

Reply:
How about `results_callback`? I would like to indicate somehow that this is a callback.

Reply:
yes that works